### ISSUE TITLE
PublisherTool stops sending messages,logs exception upon exit following suspension

### ISSUE THREAD

**Issue by antrongladney** [real-logic/Aeron#122]
Summary:
After the PublisherTool is resumed, no messages are sent,however it remains active. The PublisherTool and the MediaDriverTool log exceptions when Ctrl-C is used to shutdown the PublisherTool.

Procedures:
Start the MediaDriverTool buckwasher using the following command
java -cp tools.jar -Daeron.dir=/dev/shm/ag/aeron uk.co.real_logic.aeron.tools.MediaDriverTool

Start the SubscriberTool on buckwasher using the following commands
java -cp tools.jar -Daeron.dir=/dev/shm/ag/aeron uk.co.real_logic.aeron.tools.SubscriberTool -c 'aeron:udp?remote=192.168.14.34:31111|local=192.168.14.34'

Start the MediaDriverTool on cooper
java -cp tools.jar -Daeron.dir=/dev/shm/ag/aeron uk.co.real_logic.aeron.tools.MediaDriverTool

Start the PublisherTool on cooper
java -cp tools.jar -Daeron.dir=/dev/shm/ag/aeron uk.co.real_logic.aeron.tools.PublisherTool -c 'aeron:udp?remote=192.168.14.34:31111|local=192.168.14.31'

Allowed the subscriber to receive messages from the publisher for a few secs, before suspending the publisher via Ctrl Z. As expected the subscriber did not receive any messages while the publisher was suspended. After 5 secs the publisher was resumed. From the output of the Publisher console, it seems that the publisher sends a few more messages and then stops sending. However it remains active.
The subscriber does NOT receive any additional messages after the publisher is resumed. It reports the connection is inactive. See the output of the publisher below:

-bash-4.1$ java -cp samples.jar -Daeron.dir=/dev/shm/ag/aeron uk.co.real_logic.aeron.tools.PublisherTool -c 'aeron:udp?remote=192.168.14.34:31111|local=192.168.14.31'
INFO publisher-0 publishing 9223372036854775807 messages to: aeron:udp?remote=192.168.14.34:31111|local=192.168.14.31#1[-1682731615]
INFO Thread publisher-0 using random seed -8001883976110746674.
INFO 1.000022: 13.000 msgs/sec 3.3 Kbps
INFO 1.000034: 126.923 Kmsgs/sec 32.5 Mbps
INFO 1.000014: 282.558 Kmsgs/sec 72.3 Mbps
INFO 1.000010: 291.410 Kmsgs/sec 74.6 Mbps
INFO 0.999985: 291.010 Kmsgs/sec 74.5 Mbps
INFO 0.999946: 300.106 Kmsgs/sec 76.8 Mbps
INFO 1.000000: 292.954 Kmsgs/sec 75.0 Mbps
INFO 1.000001: 300.242 Kmsgs/sec 76.9 Mbps
INFO 0.999999: 300.195 Kmsgs/sec 76.8 Mbps
INFO 1.000001: 300.612 Kmsgs/sec 77.0 Mbps
INFO 1.000013: 300.076 Kmsgs/sec 76.8 Mbps
INFO 0.999996: 300.240 Kmsgs/sec 76.9 Mbps
INFO 1.000151: 705.268 Kmsgs/sec 180.5 Mbps
INFO 0.999838: 600.203 Kmsgs/sec 153.7 Mbps
INFO 1.000000: 579.982 Kmsgs/sec 148.5 Mbps
INFO 1.000015: 562.828 Kmsgs/sec 144.1 Mbps
INFO 1.000257: 585.655 Kmsgs/sec 149.9 Mbps
INFO 0.999859: 577.953 Kmsgs/sec 148.0 Mbps

[1]+ Stopped java -cp samples.jar -Daeron.dir=/dev/shm/ag/aeron uk.co.real_logic.aeron.tools.PublisherTool -c 'aeron:udp?remote=192.168.14.34:31111|local=192.168.14.31'
-bash-4.1$ fg
java -cp samples.jar -Daeron.dir=/dev/shm/ag/aeron uk.co.real_logic.aeron.tools.PublisherTool -c 'aeron:udp?remote=192.168.14.34:31111|local=192.168.14.31'
INFO 7.216771: 17.915 Kmsgs/sec 4.6 Mbps
INFO 0.000469: 0.000 msgs/sec 0 bps
INFO 0.000454: 0.000 msgs/sec 0 bps
INFO 0.003133: 218.343 Kmsgs/sec 55.9 Mbps
INFO 0.000362: 0.000 msgs/sec 0 bps
INFO 0.000314: 0.000 msgs/sec 0 bps
INFO 0.001093: 0.000 msgs/sec 0 bps
INFO 0.777274: 190.397 Kmsgs/sec 48.7 Mbps
INFO 1.000010: 0.000 msgs/sec 0 bps
INFO 1.000002: 0.000 msgs/sec 0 bps
INFO 0.999986: 0.000 msgs/sec 0 bps
INFO 1.000015: 0.000 msgs/sec 0 bps
INFO 0.999985: 0.000 msgs/sec 0 bps
INFO 1.000000: 0.000 msgs/sec 0 bps
INFO 1.000014: 0.000 msgs/sec 0 bps
INFO 0.999986: 0.000 msgs/sec 0 bps
INFO 0.999999: 0.000 msgs/sec 0 bps
INFO 1.000000: 0.000 msgs/sec 0 bps
INFO 1.000000: 0.000 msgs/sec 0 bps
INFO 1.000016: 0.000 msgs/sec 0 bps

While the Publisher was in this state of not sending messages, Ctrl -C was used to shutdown/exit the publisher. As a result the publisher reported the following exception and hung:
Exception in thread "publisher-0" uk.co.real_logic.aeron.exceptions.RegistrationException: Unknown publication: 1
at uk.co.real_logic.aeron.ClientConductor.onError(ClientConductor.java:258)
at uk.co.real_logic.aeron.DriverListenerAdapter.onError(DriverListenerAdapter.java:134)
at uk.co.real_logic.aeron.DriverListenerAdapter.onMessage(DriverListenerAdapter.java:119)
at uk.co.real_logic.agrona.concurrent.broadcast.CopyBroadcastReceiver.receive(CopyBroadcastReceiver.java:84)
at uk.co.real_logic.aeron.DriverListenerAdapter.receiveMessages(DriverListenerAdapter.java:56)
at uk.co.real_logic.aeron.ClientConductor.doWork(ClientConductor.java:104)
at uk.co.real_logic.agrona.concurrent.AgentRunner.run(AgentRunner.java:83)
at java.lang.Thread.run(Thread.java:745)
INFO Exiting. Sent 6976274 messages (223240768 bytes) total. 6976274 verifiable and 0 non-verifiable.

uk.co.real_logic.aeron.exceptions.DriverTimeoutException: Driver has been inactive for over 10000ms
at uk.co.real_logic.aeron.ClientConductor.checkDriverHeartbeat(ClientConductor.java:353)
at uk.co.real_logic.aeron.ClientConductor.onKeepalive(ClientConductor.java:338)
at uk.co.real_logic.aeron.ClientConductor$$Lambda$7/1555093762.run(Unknown Source)
at uk.co.real_logic.agrona.TimerWheel.expireTimers(TimerWheel.java:234)
at uk.co.real_logic.aeron.ClientConductor.processTimers(ClientConductor.java:329)
at uk.co.real_logic.aeron.ClientConductor.doWork(ClientConductor.java:103)
at uk.co.real_logic.agrona.concurrent.AgentRunner.run(AgentRunner.java:83)
at java.lang.Thread.run(Thread.java:745)

The driver also logged an exception and hung:
[671379.155471] EXCEPTION [605/605]: uk.co.real_logic.aeron.driver.exceptions.Co ntrolProtocolException(Unknown publication: 1)
uk.co.real_logic.aeron.driver.DriverConductor.onRemovePublication DriverConducto r.java:453
uk.co.real_logic.aeron.driver.DriverConductor.onClientCommand DriverConductor.ja va:288
uk.co.real_logic.aeron.driver.DriverConductor$$Lambda$35/1334729950.onMessage nu ll:-1
uk.co.real_logic.agrona.concurrent.ringbuffer.ManyToOneRingBuffer.read ManyToOne RingBuffer.java:146
uk.co.real_logic.agrona.concurrent.ringbuffer.ManyToOneRingBuffer.read ManyToOne RingBuffer.java:112

NOTE:
The same behavior is seen if the MediaDriverTool,PublisherTool, and SubscriberTool are all on the same machine

Question regarding the expected results:
Even though the client liveness timeout is 5 secs, when the publisher resumes, it is NOT a new publisher or session. Therefore, communication should continue and the connection should NOT be declared inactive. Is this correct?

**Comment by tmontgomery** [comment]
Can this be retested now with the latest set of changes?

**Comment by rlankenau** [comment]
This still happens with the latest code.  I'm working on tracking down more details.

**Comment by rlankenau** [comment]
Here's a jstack while it's hung:

2015-06-09 11:18:39
Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.31-b07 mixed mode):

"Attach Listener" #17 daemon prio=9 os_prio=31 tid=0x00007fc80910c800 nid=0x5807 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"DestroyJavaVM" #16 prio=5 os_prio=31 tid=0x00007fc80a0a4000 nid=0xe07 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"aeron-client-conductor" #12 prio=5 os_prio=31 tid=0x00007fc8098f8000 nid=0x5403 waiting on condition [0x000000012e65b000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:338)
	at uk.co.real_logic.agrona.concurrent.SleepingIdleStrategy.idle(SleepingIdleStrategy.java:41)
	at uk.co.real_logic.agrona.concurrent.AgentRunner.run(AgentRunner.java:94)
	at java.lang.Thread.run(Thread.java:745)

"Service Thread" #9 daemon prio=9 os_prio=31 tid=0x00007fc809803800 nid=0x5003 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread3" #8 daemon prio=9 os_prio=31 tid=0x00007fc809005800 nid=0x4e03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread2" #7 daemon prio=9 os_prio=31 tid=0x00007fc808843000 nid=0x4c03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #6 daemon prio=9 os_prio=31 tid=0x00007fc808842000 nid=0x4a03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=31 tid=0x00007fc808875000 nid=0x4803 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" #4 daemon prio=9 os_prio=31 tid=0x00007fc809805800 nid=0x3d0b runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=31 tid=0x00007fc809801800 nid=0x3503 in Object.wait() [0x000000012c0a8000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000076ab062f8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:142)
	- locked <0x000000076ab062f8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:158)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

"Reference Handler" #2 daemon prio=10 os_prio=31 tid=0x00007fc80886d800 nid=0x3303 in Object.wait() [0x000000012bfa5000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000076ab05d68> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:157)
	- locked <0x000000076ab05d68> (a java.lang.ref.Reference$Lock)

"VM Thread" os_prio=31 tid=0x00007fc80886b000 nid=0x3103 runnable 

"GC task thread#0 (ParallelGC)" os_prio=31 tid=0x00007fc808818800 nid=0x1907 runnable 

"GC task thread#1 (ParallelGC)" os_prio=31 tid=0x00007fc808819000 nid=0x2313 runnable 

"GC task thread#2 (ParallelGC)" os_prio=31 tid=0x00007fc80881a000 nid=0x2503 runnable 

"GC task thread#3 (ParallelGC)" os_prio=31 tid=0x00007fc80881a800 nid=0x2703 runnable 

"GC task thread#4 (ParallelGC)" os_prio=31 tid=0x00007fc80881b000 nid=0x2903 runnable 

"GC task thread#5 (ParallelGC)" os_prio=31 tid=0x00007fc80881b800 nid=0x2b03 runnable 

"GC task thread#6 (ParallelGC)" os_prio=31 tid=0x00007fc80881c800 nid=0x2d03 runnable 

"GC task thread#7 (ParallelGC)" os_prio=31 tid=0x00007fc80881d000 nid=0x2f03 runnable 

"VM Periodic Task Thread" os_prio=31 tid=0x00007fc808843800 nid=0x5203 waiting on condition 

JNI global references: 338

**Comment by tmontgomery** [comment]
I'm not seeing any hung threads here. Or I should say, any threads that are waiting on stuff from Aeron for whatever reason.

I think what is happening is that the publication goes inactive since the client (publisher) doesn't show liveness. It therefore is reclaimed by the media driver as intended. When it comes back, it continues to publish, but the media driver isn't paying attention.

The publisher is still able to publish, though, into the logbuffer. But it isn't going anywhere.

In short, the client (publishers and receivers) have to show liveness or they will be timed out. If the timeout is too short, then it can be extended via configuration.

One additional check, when the publisher is subsequently closed, does the media driver still work?

**Comment by mjpt777** [comment]
Closing due to no further information.
